File: 07_linked_list/07_11_intersection_of_two_linked_lists/intersection_of_two_linked_lists.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def getIntersectionNode(headA, headB):
    if not headA or not headB:
        return None
    a, b = headA, headB
    while a != b:
        a = headB if a is None else a.next
        b = headA if b is None else b.next
    return a

def test_intersection(label, headA, headB, expected):
    print(label, end=" ")
    intersection_node = getIntersectionNode(headA, headB)
    print_list(intersection_node)
    print(f" (Expected: {expected.val if expected else 'None'})")

def print_list(head):
    current = head
    while current:
        print(current.val, "-> ", end="")
        current = current.next
    print("None")

# Test Case 1: Intersecting at node with value 8
common_node1 = ListNode(8, ListNode(4, ListNode(5)))
headA1 = ListNode(4, ListNode(1, common_node1))
headB1 = ListNode(5, ListNode(6, ListNode(1, common_node1)))
test_intersection("Test Case 1:", headA1, headB1, common_node1)  # Output: 8 -> 4 -> 5

# Test Case 2: No intersection
headA2 = ListNode(2, ListNode(6, ListNode(4)))
headB2 = ListNode(1, ListNode(5))
test_intersection("Test Case 2:", headA2, headB2, None)  # Output: None

# Test Case 3: Intersecting at node with value 2
common_node3 = ListNode(2, ListNode(4, ListNode(5)))
headA3 = ListNode(1, ListNode(9, ListNode(1, common_node3)))
headB3 = ListNode(3, common_node3)
test_intersection("Test Case 3:", headA3, headB3, common_node3)  # Output: 2 -> 4 -> 5

File: 07_linked_list/07_11_intersection_of_two_linked_lists/intersection_of_two_linked_lists.cpp
#include <iostream>

class ListNode {
public:
    int val;
    ListNode* next;

    ListNode(int value = 0, ListNode* nextNode = nullptr) : val(value), next(nextNode) {}
};

ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
    if (!headA || !headB) {
        return nullptr;
    }
    ListNode* a = headA;
    ListNode* b = headB;
    while (a != b) {
        a = a ? a->next : headB;
        b = b ? b->next : headA;
    }
    return a;
}

void printList(ListNode* head) {
    ListNode* current = head;
    while (current) {
        std::cout << current->val << " -> ";
        current = current->next;
    }
    std::cout << "null";
}

void testIntersection(const std::string& label, ListNode* headA, ListNode* headB, ListNode* expected) {
    std::cout << label << " ";
    printList(getIntersectionNode(headA, headB));
    std::cout << "(Expected: " << (expected ? std::to_string(expected->val) : "null") << ")" << std::endl;
}

int main() {
    // Test Case 1: Intersecting at node with value 8
    ListNode* commonNode1 = new ListNode(8, new ListNode(4, new ListNode(5)));
    ListNode* headA1 = new ListNode(4, new ListNode(1, commonNode1));
    ListNode* headB1 = new ListNode(5, new ListNode(6, new ListNode(1, commonNode1)));
    testIntersection("Test Case 1:", headA1, headB1, commonNode1);  // Output: 8 -> 4 -> 5

    // Test Case 2: No intersection
    ListNode* headA2 = new ListNode(2, new ListNode(6, new ListNode(4)));
    ListNode* headB2 = new ListNode(1, new ListNode(5));
    testIntersection("Test Case 2:", headA2, headB2, nullptr);  // Output: null

    // Test Case 3: Intersecting at node with value 2
    ListNode* commonNode3 = new ListNode(2, new ListNode(4, new ListNode(5)));
    ListNode* headA3 = new ListNode(1, new ListNode(9, new ListNode(1, commonNode3)));
    ListNode* headB3 = new ListNode(3, commonNode3);
    testIntersection("Test Case 3:", headA3, headB3, commonNode3);  // Output: 2 -> 4 -> 5

    // Remember to free allocated memory
    delete commonNode1;
    delete headA1;
    delete headB1;
    delete headA2;
    delete headB2;
    delete commonNode3;
    delete headA3;
    delete headB3;

    return 0;
}
